### PR TITLE
SNAPSHOT: Keep SnapshotsInProgress State in Sync with Routing Table

### PR DESCRIPTION
**Pull request by original-brownbear**
* Keep `SnapshotsInProgress` state in sync with routing table
* Handle being unable to get shard from indicesService gracefully
* Add assertions about inner consistency of `SnapShotsInProgress` with routing table

### CODE REVIEW

**Comment by original-brownbear** [comment]
@ywelsch 

I think this is good for a review if you have some time.
I organized the signature of `updateWithRoutingTable` and the assertion on the state a little differently than in your initial approach. I like this better because you get an assertion on the full `SnapshotsInProgress` state instead of failing right on the first shard entry that is off => you get a little more information out of the assert message. Also, I found `updateWithRoutingTable` a litlle easier to digest this way, but let me know what you think :)

**Comment by original-brownbear** [comment]
@ywelsch looks like it's green now :) 

I had to relax the overall state assertion a little though and not run it when the master was just elected (here https://github.com/elastic/elasticsearch/pull/35710/files#diff-a0853be4492c052f24917b5c1464003dR662)

This is a result of the cleanup action here https://github.com/elastic/elasticsearch/pull/35710/files#diff-a0853be4492c052f24917b5c1464003dR654 not happening in this case as far as I can tell. You refactored that logic too in your branch and it's fine there but I guess we can leave that for the next round?

(PS: Also ran `:server:test :server:integTest` in a loop locally for a few hours today without issues with this code :))

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 1]
```diff
@@ -137,14 +138,28 @@ private ClusterState buildResult(ClusterState oldState, RoutingAllocation alloca
             .routingTable(newRoutingTable)
             .metaData(newMetaData);
         final RestoreInProgress restoreInProgress = allocation.custom(RestoreInProgress.TYPE);
+        ImmutableOpenMap.Builder<String, ClusterState.Custom> customsBuilder = null;
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: move this line above the one with `final RestoreInProgress restoreInProgress = allocation.custom(RestoreInProgress.TYPE);`?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java [hunk 2]
```diff
@@ -76,11 +78,12 @@
     private final IndexMetaDataUpdater indexMetaDataUpdater = new IndexMetaDataUpdater();
     private final RoutingNodesChangedObserver nodesChangedObserver = new RoutingNodesChangedObserver();
     private final RestoreInProgressUpdater restoreInProgressUpdater = new RestoreInProgressUpdater();
+    private final SnapshotsService.SnapshotsInProgressUpdater snapshotsInProgressUpdater =
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java]: nit: statically import SnapshotsInProgressUpdater

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java [hunk 3]
```diff
@@ -328,7 +328,11 @@ private void processIndexShardSnapshots(ClusterChangedEvent event) {
 
                 for (final Map.Entry<ShardId, IndexShardSnapshotStatus> shardEntry : entry.getValue().entrySet()) {
                     final ShardId shardId = shardEntry.getKey();
-                    final IndexShard indexShard = indicesService.indexServiceSafe(shardId.getIndex()).getShardOrNull(shardId.id());
+                    final IndexShard indexShard = indicesService.getShardOrNull(shardId);
+                    if (indexShard == null) {
+                        notifyFailedSnapshotShard(snapshot, shardId, localNodeId, "could not find shard");
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java]: Is this change necessary to fix the test failure? Should we do this as part of the follow-up refactoring of SnapshotShardsService?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 4]
```diff
@@ -657,12 +659,28 @@ public void applyClusterState(ClusterChangedEvent event) {
                 }
                 removeFinishedSnapshotFromClusterState(event);
                 finalizeSnapshotDeletionFromPreviousMaster(event);
+                // TODO org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT.testDeleteOrphanSnapshot fails right after election here
+                assert event.previousState().nodes().isLocalNodeElectedMaster() != false || assertConsistency(event.state());
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: `X != false <==> X`

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: Note that we'll have a similar problem when transitioning from a master that did not keep the routing table in sync to a master that now does. We will probably have to schedule a clean-up task that brings these two in sync again. There's no guarantee that that task will run before any other tasks, so the assertion might not only be violated while `event.previousState().nodes().isLocalNodeElectedMaster() == false`, but also on some follow-up cluster state changes. I'm wondering if we need to introduce some state to SnapshotsService to say whether we've properly cleaned up as a master and then make that assertion dependent on that boolean variable. Something we can look into in a follow-up.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 5]
```diff
@@ -1539,6 +1557,126 @@ public RepositoriesService getRepositoriesService() {
...
+        if (oldSnapshot == null || shardIds.isEmpty()) {
+            return oldSnapshot;
+        }
+        List<SnapshotsInProgress.Entry> entries = new ArrayList<>();
+        boolean snapshotsInProgressChanged = false;
+        for (SnapshotsInProgress.Entry entry : oldSnapshot.entries()) {
+            ImmutableOpenMap.Builder<ShardId, ShardSnapshotStatus> shardsBuilder = null;
+            for (ShardId shardId : shardIds) {
+                ImmutableOpenMap<ShardId, ShardSnapshotStatus> shards = entry.shards();
+                ShardSnapshotStatus currentStatus = shards.get(shardId);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: make these two variables final?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 6]
```diff
@@ -1539,6 +1557,126 @@ public RepositoriesService getRepositoriesService() {
...
+                if (currentStatus != null && currentStatus.state().completed() == false) {
+                    final ShardSnapshotStatus newStatus = Optional
+                        .ofNullable(newRoutingTable.shardRoutingTableOrNull(shardId))
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: I wonder if we can expect there to always exist the corresponding shard routing table?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 7]
```diff
@@ -1539,6 +1557,126 @@ public RepositoriesService getRepositoriesService() {
...
+            return new SnapshotsInProgress(entries);
+        }
+        return oldSnapshot;
+    }
+
+    private static ShardSnapshotStatus determineShardSnapshotStatus(final ShardSnapshotStatus currentStatus,
+        final ShardRouting primaryShardRouting) {
+        final State currentState = currentStatus.state();
+        final ShardSnapshotStatus newStatus;
+        if (primaryShardRouting == null) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: can this ever be null here?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 8]
```diff
@@ -1539,6 +1557,126 @@ public RepositoriesService getRepositoriesService() {
...
+                        newStatus = currentStatus;
+                    } else {
+                        newStatus = failedStatus(currentNodeId);
+                    }
+                    break;
+                default:
+                    newStatus = currentStatus;
+                    break;
+            }
+        } else if (currentState == State.INIT || currentStatus.state() == State.ABORTED) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: can you add a comment that we are in the relocating state here

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left mostly smaller comments on the PR. Can you also rename the PR to better reflect what's been done here, namely to keep SnapshotsInProgress state in sync with routing table?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java]: @ywelsch I was able to reproduce running into this one with a very low frequency running the newly added test (though I just tried it again for an hour in a loop and couldn't so maybe this isn't an issue anymore ... I'll revert it for now, technically this shouldn't be necessary).

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 7 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: Oh no, my bad for not cleaning that up, the caller ensures this is never `null` :)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 6 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: I don't know but can't a `ShardId` that we got from the callback `org.elasticsearch.cluster.routing.RoutingChangesObserver.AbstractChangedShardObserver#shardFailed` lead to a `null` entry here? (I could be way off here ... not sure about the order of these things yet)

**Comment by original-brownbear** [comment]
@ywelsch all points addressed I think :)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 8 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: Added the assertion for that below.

**Comment by original-brownbear** [comment]
@ywelsch ping (if you have a second) :)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 6 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: we are never adding or removing keys in the shard routing table during the `AllocationService.reroute()` phase. Can you add an assertion that the routing table is not null?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 8 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: perhaps move the assertion one level up, i.e.,

```
} else {
  assert priamryShardRouting.relocating();
  if (currentState == State.INIT || currentStatus.state() == State.ABORTED) { 
    ..
  } else {
    ..
  }
}
```

**Review by ywelsch — APPROVED** [review]
2 nits, looks good otherwise.

**Comment by original-brownbear** [comment]
@ywelsch thanks!, nits handled => merging :)

**Pull request merged**